router/broker.go

```diff
@@ -305,47 +362,80 @@ func (b *broker) syncPublish(pub *wamp.Session, msg *wamp.Publish, pubID wamp.ID
 }
 
 func newSubscription(id wamp.ID, subscriber *wamp.Session, topic wamp.URI, match string) *subscription {
+	subscribers := map[*wamp.Session]struct{}{}
+	if subscriber != nil {
+		subscribers = map[*wamp.Session]struct{}{subscriber: {}}
+	}
+
 	return &subscription{
 		id:          id,
 		topic:       topic,
 		match:       match,
 		created:     wamp.NowISO8601(),
-		subscribers: map[*wamp.Session]struct{}{subscriber: {}},
+		subscribers: subscribers,
 	}
 }
 
-func (b *broker) syncSubscribe(subscriber *wamp.Session, msg *wamp.Subscribe, match string) {
+func (b *broker) syncSaveEvent(eventStore *eventHistoryStoreItem, pub *wamp.Publish, event *wamp.Event) {
+	item := &eventHistoryItem{
+		publication: pub,
+		event: &storedEvent{
+			timestamp:    time.Now(),
+			Subscription: event.Subscription,
+			Publication:  event.Publication,
+			Details:      event.Details,
+			Arguments:    event.Arguments,
+			ArgumentsKw:  event.ArgumentsKw,
+		},
+	}
+	eventStore.events = append(eventStore.events, item)
```

Comment:
I'd move this part after the re-slicing below and move the condition:

```
if len(eventStore.events) >= eventStore.limit {
   eventStore.isLimitReached = true
   eventStore.events = ... 
}
```

Then below you know for a fact that you have space and can just store/append/etc.